### DISCUSSION TITLE
GitHub Web Hooks documentation

### DISCUSSION THREAD

**Discussion by VladMasarik** [community/community#24312] [Apps, API and Webhooks]
<p>Where can I find documentation to triggering webhooks when a comment is created on a PR? Also, where is it explained that a PR is just an issue with code?</p>
<p>I just finished 3 hours long search for how to send a GitHub webhook when a comment is created on a PR. I believe I searched the Webhooks documentation through and through, and nowhere, is any mention about the fact that comments on a PR actually trigger the <code>issue_comment</code> trigger. I had to randomly stumble uppon <a>this</a>.</p>
<p>I believe there has to be more information about that fact than that short sentence. There is no way you mention that “a Pull Request is just an Issue with code” only once. Mainly, if the documentation itself states “This may seem counterintuitive at first.”.</p>
<p>Also, sorry for being too hostile, but I feel exhasted after searching for such a simple thing for so long, and the authors themselves acknowledged that it is counterintuitive. Yet, they never explained the reasons for it, nor was that information easy to find.</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hey <a class="mention" href="/vladmasarik">@vladmasarik</a>,</p>
<p>Thanks for being here, and thanks for taking the time to write this feedback. I can see how is very frustrating to not be able to find clear documentation. Let me see if I can help clarify some of your questions and rest assured I’ll pass on your feedback to our docs team.</p>
<p>Here is a link to the current docs for our Webhooks product (<a href="https://developer.github.com/webhooks/" rel="nofollow noopener">https://developer.github.com/webhooks/</a>). These are configured at repository level and the principal is that you can subscribe to events that happen in the given repository and receive a payload representing this, which you can then process as you see fit, storing the data locally or triggering some desired behavior.</p>
<p>When you create a hook you designate a URL for this hook. This has to be a publicly exposed endpoint that can receive POST requests. Once you have this you can then configure your hook to either receive form-encoded POST data, or JSON encoded POST data. Finally, you pick the events that you’d like to subscribe to this hook to.</p>
<p>When an event occurs, such as a push or triggered when a pull request is <code>assigned</code>, <code>unassigned</code>, <code>labeled</code>, <code>unlabeled</code>, <code>opened</code>, <code>edited</code>, <code>closed</code>, <code>reopened</code>, <code>synchronize</code>, <code>ready_for_review</code>, <code>locked</code>, <code>unlocked</code> we will check the hooks to see if there are any that are subscribed to this event. If there is, we will generate a payload and through a job on our side trigger a POST request with this payload to your endpoint. The payload will look something like:</p>
<p><a href="https://developer.github.com/v3/activity/events/types/#pushevent" rel="nofollow noopener">https://developer.github.com/v3/activity/events/types/#pushevent</a></p>
<pre><code>{
  "ref": "refs/tags/simple-tag",
  "before": "a10867b14bb761a232cd80139fbd4c0d33264240",
  "after": "0000000000000000000000000000000000000000",
  "created": false,
  "deleted": true,
  "forced": false,
  "base_ref": null,
  "compare": "https://github.com/Codertocat/Hello-World/compare/a10867b14bb7...000000000000",
  "commits": [

  ],
  "head_commit": null,
  "repository": {
    "id": 135493233,
    "node_id": "MDEwOlJlcG9zaXRvcnkxMzU0OTMyMzM=",
    "name": "Hello-World",
    "full_name": "Codertocat/Hello-World",
    "owner": {
      "name": "Codertocat",
      "email": "21031067+Codertocat@users.noreply.github.com",
      "login": "Codertocat",
      "id": 21031067,
      "node_id": "MDQ6VXNlcjIxMDMxMDY3",
      "avatar_url": "https://avatars1.githubusercontent.com/u/21031067?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/Codertocat",
      "html_url": "https://github.com/Codertocat",
      "followers_url": "https://api.github.com/users/Codertocat/followers",
      "following_url": "https://api.github.com/users/Codertocat/following{/other_user}",
      "gists_url": "https://api.github.com/users/Codertocat/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/Codertocat/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/Codertocat/subscriptions",
      "organizations_url": "https://api.github.com/users/Codertocat/orgs",
      "repos_url": "https://api.github.com/users/Codertocat/repos",
      "events_url": "https://api.github.com/users/Codertocat/events{/privacy}",
      "received_events_url": "https://api.github.com/users/Codertocat/received_events",
      "type": "User",
      "site_admin": false
    },
    "private": false,
    "html_url": "https://github.com/Codertocat/Hello-World",
    "description": null,
    "fork": false,
    "url": "https://github.com/Codertocat/Hello-World",
    "forks_url": "https://api.github.com/repos/Codertocat/Hello-World/forks",
    "keys_url": "https://api.github.com/repos/Codertocat/Hello-World/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/Codertocat/Hello-World/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/Codertocat/Hello-World/teams",
    "hooks_url": "https://api.github.com/repos/Codertocat/Hello-World/hooks",
    "issue_events_url": "https://api.github.com/repos/Codertocat/Hello-World/issues/events{/number}",
    "events_url": "https://api.github.com/repos/Codertocat/Hello-World/events",
    "assignees_url": "https://api.github.com/repos/Codertocat/Hello-World/assignees{/user}",
    "branches_url": "https://api.github.com/repos/Codertocat/Hello-World/branches{/branch}",
    "tags_url": "https://api.github.com/repos/Codertocat/Hello-World/tags",
    "blobs_url": "https://api.github.com/repos/Codertocat/Hello-World/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/Codertocat/Hello-World/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/Codertocat/Hello-World/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/Codertocat/Hello-World/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/Codertocat/Hello-World/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/Codertocat/Hello-World/languages",
    "stargazers_url": "https://api.github.com/repos/Codertocat/Hello-World/stargazers",
    "contributors_url": "https://api.github.com/repos/Codertocat/Hello-World/contributors",
    "subscribers_url": "https://api.github.com/repos/Codertocat/Hello-World/subscribers",
    "subscription_url": "https://api.github.com/repos/Codertocat/Hello-World/subscription",
    "commits_url": "https://api.github.com/repos/Codertocat/Hello-World/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/Codertocat/Hello-World/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/Codertocat/Hello-World/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/Codertocat/Hello-World/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/Codertocat/Hello-World/contents/{+path}",
    "compare_url": "https://api.github.com/repos/Codertocat/Hello-World/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/Codertocat/Hello-World/merges",
    "archive_url": "https://api.github.com/repos/Codertocat/Hello-World/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/Codertocat/Hello-World/downloads",
    "issues_url": "https://api.github.com/repos/Codertocat/Hello-World/issues{/number}",
    "pulls_url": "https://api.github.com/repos/Codertocat/Hello-World/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/Codertocat/Hello-World/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/Codertocat/Hello-World/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/Codertocat/Hello-World/labels{/name}",
    "releases_url": "https://api.github.com/repos/Codertocat/Hello-World/releases{/id}",
    "deployments_url": "https://api.github.com/repos/Codertocat/Hello-World/deployments",
    "created_at": 1527711484,
    "updated_at": "2018-05-30T20:18:35Z",
    "pushed_at": 1527711528,
    "git_url": "git://github.com/Codertocat/Hello-World.git",
    "ssh_url": "git@github.com:Codertocat/Hello-World.git",
    "clone_url": "https://github.com/Codertocat/Hello-World.git",
    "svn_url": "https://github.com/Codertocat/Hello-World",
    "homepage": null,
    "size": 0,
    "stargazers_count": 0,
    "watchers_count": 0,
    "language": null,
    "has_issues": true,
    "has_projects": true,
    "has_downloads": true,
    "has_wiki": true,
    "has_pages": true,
    "forks_count": 0,
    "mirror_url": null,
    "archived": false,
    "open_issues_count": 2,
    "license": null,
    "forks": 0,
    "open_issues": 2,
    "watchers": 0,
    "default_branch": "master",
    "stargazers": 0,
    "master_branch": "master"
  },
  "pusher": {
    "name": "Codertocat",
    "email": "21031067+Codertocat@users.noreply.github.com"
  },
  "sender": {
    "login": "Codertocat",
    "id": 21031067,
    "node_id": "MDQ6VXNlcjIxMDMxMDY3",
    "avatar_url": "https://avatars1.githubusercontent.com/u/21031067?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/Codertocat",
    "html_url": "https://github.com/Codertocat",
    "followers_url": "https://api.github.com/users/Codertocat/followers",
    "following_url": "https://api.github.com/users/Codertocat/following{/other_user}",
    "gists_url": "https://api.github.com/users/Codertocat/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/Codertocat/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/Codertocat/subscriptions",
    "organizations_url": "https://api.github.com/users/Codertocat/orgs",
    "repos_url": "https://api.github.com/users/Codertocat/repos",
    "events_url": "https://api.github.com/users/Codertocat/events{/privacy}",
    "received_events_url": "https://api.github.com/users/Codertocat/received_events",
    "type": "User",
    "site_admin": false
  }
}
</code></pre>
<p>If the event has multiple actions, for instance an issue event might have an event for created, updated, a new label being added etc. Then there is an action key in this payload that denotes what action triggered this event.</p>
<p>Once you receive this it’s up to you how you decide to process this, for instance, you might store this in a db or key/value no-db store, you might decide to store all of the data, or just what you need for your personal site - the choice is entirely yours.</p>
<p>I hope this helps, and if you have any further questions feel free to reach out.</p>

**Comment by okainov** [comment]
<p>Thanks for so useless answer. This doesn’t help even a bit. I think you misread the OP’s problem completely.</p>
<ol>
<li>Implementation IS full of workarounds which you you try to justify by saying " Pull Request is just an Issue with code", but in reality we all know that’s not true and it would be so much easier if Github team acknowledge that and implement proper endpoints and webhooks for both.</li>
<li>Documentation IS unclear. It’s even written there “This may seem counterintuitive”, quote.</li>
<li>Implementation IS inconsistent. Pull request data in “pull_request” events payload is completely different than in “issue_comment” (which is not an issue_comment but in this context pull_request_comment). IN particular, issue_comment payload doesn’t have information about pull request at all. The ID of PR is on different JSON levels, one is on root level, another one is on the second level.</li>
</ol>

**Reply by brandonzylstra** [reply]
@okainov 

> Thanks for so useless answer. This doesn’t help even a bit.

A little patience with @AndreaGriffiths11 and less sarcasm would be helpful if you want to solve the problems you've described.  Please keep in mind that _**she**_ is not responsible for the current mess with the way GitHub handles PRs being issues.  (Apparently they evolved out of issues, where people would simply open an issue to request that someone pull their branch into the master branch, and new features were added to them over time but GitHub never made PRs into first class objects even after they evolved into something completely different from issues.)  Nor is she responsible for the current state of the documentation.  She's just trying to help  @VladMasarik in any way she can, after him having run into these problems.

She was very gracious despite the circumstances.  Read her first sentence to see what I mean:

> Thanks for being here, and thanks for taking the time to write this feedback. I can see how is very frustrating to not be able to find clear documentation. Let me see if I can help clarify some of your questions and rest assured I’ll pass on your feedback to our docs team.

But back to your comments.

> I think you misread the OP’s problem completely.

That is probably true, but how much of what she missed was explicitly stated?  Do you think the responsibility is 100% on her shoulders for understanding him?  He wrote

> Where can I find documentation to triggering webhooks when a comment is created on a PR? 

First of all, the sentence doesn't actually work with "to".  I think he means "about" or "on".  It's harder to understand anyone when you must do mental grammar correction (and _guess_ what they intended).  I'm _not_ saying that to criticize @VladMasarik.  I know English can be hard to master, and the way each language handles preposition tends to be very different, and this makes it very hard to get them right in any second language. We need to cut all non-native speakers a lot of slack since they are doing a lot of extra work to communicate in a language they didn't learn as infants. I only point this out to show that it gives @AndreaGriffiths11 extra work too on the receiving end of the message.  She's already doing extra work to communicate in English, but on top of that she is receiving English with syntax errors, and she's expected to correctly parse it.  But this is the smaller issue.

The much bigger issue with this sentence is that it's unclear whether it's a real question or a rhetorical question.  Is he saying this because he wants an answer?  Or to express how hard it is to find the documentation?  I think it's a rhetorical question, since he follows it with this question 

> Also, where is it explained that a PR is just an issue with code?

He knows that a PR is "just an issue with code", and asking where this is documented seems to be meant to show how bad the documentation is.  I _think_ that's the case, but I'm no 100% sure if that's his point.  

But my point to you is: _**is it really fair to complain that she didn't understand something that is so unclear?**_

If his question was meant to imply his point, and she interpreted it as a straightforward request for information, is that a reason to make sarcastic remarks at her? 

I don't think so.  

I think you're _probably right_ about the problems with documentation and the strangeness around PRs on GitHub.  (I haven't read the documentation about this so I can't say with certainty.)  But shooting the messenger doesn't help solve the problem.

Note that she left the conversation open, leaving opportunity for clarifications if they were needed:

> I hope this helps, and if you have any further questions feel free to reach out.

When there is misunderstanding in communication, the best next step is to try to bring clarity, not to snipe at the person who innocently misunderstood something.
